Ceph: Filesystem Pool Size are ignored

rajatsingh25aug opened the pull request:
Changing the metadataPool and dataPools replicated size no effect.

Signed-off-by: RAJAT SINGH <rajasing@redhat.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**
The old Filesystem's pool is not getting updated when changed with the new File system.
So, we added two helper functions that will generate the names of the pools and then change those pool's values using the `SetPoolSize` method.

**Which issue is resolved by this Pull Request:**
Resolves #3144 

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test ceph]
[test full]

rajatsingh25aug commented:
cc @travisn  @sp98 closed earlier PR as I messed it up while rebasing.

rajatsingh25aug commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -173,6 +174,11 @@ func IsMultiFSEnabled() bool {
 	return false
 }
+func SetPools(context *clusterd.Context, cephVersion cephver.CephVersion, clusterName, fsName string, metdataPool []*model.Pool, dataPool *model.Pool) error {
+	args := []string{"osd", "pool", "set", fsName, "size"}
```
So this is the command that is needed to update the pool size, the only problem that I'm getting is that I'm not able to use `len(metadataPool)`, I know it's due to the data type mismatch.
@travisn can you please add some suggestions?.

ashishranjan738 replied:
hope this helps https://gobyexample.com/arrays @rajatsingh25aug

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -173,6 +174,11 @@ func IsMultiFSEnabled() bool {
 	}
 	return false
 }
+func SetPools(context *clusterd.Context, cephVersion cephver.CephVersion, clusterName, fsName string, metdataPool []*model.Pool, dataPool *model.Pool) error {
```
Does this function really need to be exported?

rajatsingh25aug replied:
This function is defined here and used here https://github.com/rook/rook/pull/4433#discussion_r354334380

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -173,6 +174,11 @@ func IsMultiFSEnabled() bool {
 	return false
 }
+func SetPools(context *clusterd.Context, cephVersion cephver.CephVersion, clusterName, fsName string, metdataPool []*model.Pool, dataPool *model.Pool) error {
+	args := []string{"osd", "pool", "set", fsName, "size"}
```
I think you're missing an argument, which is the desired size.

rajatsingh25aug replied:
Yes, and it'll be the DataPool and the MetadataPool Replicated size, I'm having trouble with the data type mismatch.
`size (metadataPool)`  throws an error that it cannot convert to int.
Wondering what dataType would it be.

leseb replied:
data type of? The CLI should be:

```go
args := []string{"osd", "pool", "set", fsName, "size", replicaCount}
```

`replicaCount` is a string.

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -179,6 +178,10 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
+		if err := client.SetPools(context, cephVersion, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+
+			return nil
+		}
```
Here I'll add condition if the update is not possible, I'll take references from this function https://github.com/rook/rook/blob/ab69d75cf2f5c6279b3f044493bf32e3c3a9c779/pkg/operator/ceph/file/filesystem.go#L174 to write mine.

leseb replied:
Ok.

rajatsingh25aug commented:
ok i’ll try that

On Sat, 7 Dec 2019 at 1:45 AM, Ashish Ranjan <notifications@github.com>
wrote:

> *@ashishranjan738* commented on this pull request.
> ------------------------------
>
> In pkg/daemon/ceph/client/filesystem.go
> <https://github.com/rook/rook/pull/4433#discussion_r355012360>:
>
> > @@ -173,6 +174,11 @@ func IsMultiFSEnabled() bool {
>  	}
>  	return false
>  }
> +func SetPools(context *clusterd.Context, cephVersion cephver.CephVersion, clusterName, fsName string, metdataPool []*model.Pool, dataPool *model.Pool) error {
> +	args := []string{"osd", "pool", "set", fsName, "size"}
>
> hope this helps https://gobyexample.com/arrays @rajatsingh25aug
> <https://github.com/rajatsingh25aug>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rook/rook/pull/4433?email_source=notifications&email_token=AEDUM2CUZJA5TPJ4GPQ4UNDQXKXERA5CNFSM4JV2A3C2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOJYHBI#discussion_r355012360>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDUM2GVJZW5MMS75MFEB4DQXKXERANCNFSM4JV2A3CQ>
> .
>

rajatsingh25aug commented:
@travisn please review, still WIP.

rajatsingh25aug commented on cluster/examples/kubernetes/ceph/operator-openshift.yaml:
```diff
@@ -111,7 +111,7 @@ spec:
       containers:
       - name: rook-ceph-operator
-        image: rook/ceph:master
+        image: rajasing/ceph:test
```
yep, It’s still in WIP, I just wanted to get yours reviews.
Thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -180,6 +179,13 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 		}
+		if err := client.SetPools(context, cephVersion, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			logger.Errorf(
+				fmt.Sprintf("failed to change replicated size for dataPools to %d for filesystem %s, and replicated size for metadataPool to %d for filesystem %s still continuing. ", f.dataPools[0].ReplicatedConfig.Size, f.Name, f.metadataPool.ReplicatedConfig.Size, f.Name) +
```
I tried taking the reference from the above function, so I tried to incorporate as much info as possible.
I’ll make changes as requested.

ashishranjan738 replied:
why there is a hardcoded index with no length checks? A hardcoded index can throw IndexOutOfBound error in runtime.

rajatsingh25aug replied:
Thanks for the tip!
Added a `for` loop instead.

ashishranjan738 commented on cluster/examples/kubernetes/ceph/filesystem.yaml:
```diff
@@ -13,12 +13,12 @@ spec:
   metadataPool:
     replicated:
-      size: 3
+      size: 1
```
why this change is needed ?

rajatsingh25aug replied:
I have reverted the changes made.

rajatsingh25aug commented on pkg/operator/ceph/file/controller.go:
```diff
@@ -225,7 +225,19 @@ func (c *FilesystemController) filesystemOwner(fs *cephv1.CephFilesystem) metav1
+	if len(oldFS.DataPools) == len(newFS.DataPools) {
+		for i := 0; i < len(oldFS.DataPools); i++ {
+			if oldFS.DataPools[i].Replicated.Size != newFS.DataPools[i].Replicated.Size {
+				logger.Infof("existing DataPools replicated size changed from %d to %d", oldFS.DataPools[i].Replicated.Size, newFS.DataPools[i].Replicated.Size)
```
This is done in case multiple  `dataPools` have  been added and now the operator can detect exactly which dataPool's replicated size has been changed.
Thanks for the tip @ashishranjan738

ashishranjan738 commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -174,6 +177,23 @@ func IsMultiFSEnabled() bool {
+func SetPoolSize(context *clusterd.Context, clusterName, fsName string, dataPool []*model.Pool, metdataPool *model.Pool) error {
+	// updating the dataPool size
+	metaDataPoolname := fsName + metaDataPoolSuffix
+	err := SetPoolProperty(context, clusterName, metaDataPoolname, "size", strconv.Itoa(int(metdataPool.ReplicatedConfig.Size)))
```
How about making the property name as constants?

ashishranjan738 replied:
can be done in a separate PR.

rajatsingh25aug replied:
Because of the `fsName` is declared within the func scope and that's why I'm concatenating it with the other const that I defined above in this scope.

ashishranjan738 replied:
I mean `size` which is getting passed to SetPoolProperty.

rajatsingh25aug replied:
and there are two functions called one after the another so I think instead of "resetting" the strings, we can bear the cost of one more variable?.

ashishranjan738 commented on pkg/operator/ceph/file/controller.go:
```diff
@@ -225,7 +225,19 @@ func (c *FilesystemController) filesystemOwner(fs *cephv1.CephFilesystem) metav1
+	if len(oldFS.DataPools) == len(newFS.DataPools) {
+		for i := 0; i < len(oldFS.DataPools); i++ {
+			if oldFS.DataPools[i].Replicated.Size != newFS.DataPools[i].Replicated.Size {
+				logger.Infof("existing DataPools replicated size changed from %d to %d", oldFS.DataPools[i].Replicated.Size, newFS.DataPools[i].Replicated.Size)
```
can we also print the name of the datapool whose size got changed ?

rajatsingh25aug replied:
After I checked, there is no property for datapool's name under `FilesystemSpec struct`

ashishranjan738 commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -180,6 +179,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 					fmt.Sprintf(": %+v", err),
 			)
 		}
+		if err := client.SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
```
no need of semicolon here.

ashishranjan738 replied:
okay I think I misread it because of the split view of github. Seems okay to me.

rajatsingh25aug replied:
I think it is :)

ashishranjan738 commented:
Overall logic looks good to me. Approving it.

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -174,6 +177,23 @@ func IsMultiFSEnabled() bool {
+	metaDataPoolname := fsName + metaDataPoolSuffix
+	err := SetPoolProperty(context, clusterName, metaDataPoolname, "size", strconv.Itoa(int(metdataPool.ReplicatedConfig.Size)))
+	if err != nil {
+		return errors.Wrapf(err, "failed to update metadataPool replica for filesystem %s", fsName)
```
```suggestion
		return errors.Wrapf(err, "failed to update metadataPool replica for filesystem %q", fsName)
```

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -174,6 +177,23 @@ func IsMultiFSEnabled() bool {
+		err = SetPoolProperty(context, clusterName, dataPoolname+strconv.Itoa(int(i)), "size", strconv.Itoa(int(dataPool[i].ReplicatedConfig.Size)))
+	}
+	if err != nil {
+		return errors.Wrapf(err, "failed to update dataPool replica for filesystem %s", fsName)
```
```suggestion
		return errors.Wrapf(err, "failed to update dataPool replica for filesystem %q", fsName)
```

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -174,6 +177,23 @@ func IsMultiFSEnabled() bool {
+	//updating dataPool size
+	dataPoolname := fsName + dataPoolSuffix
+	for i := 0; i < len(dataPool); i++ {
+		err = SetPoolProperty(context, clusterName, dataPoolname+strconv.Itoa(int(i)), "size", strconv.Itoa(int(dataPool[i].ReplicatedConfig.Size)))
```
You should move the `if err != nil {` in the loop.

rajatsingh25aug replied:
My bad, I didn't see that!

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -180,6 +179,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 			)
 		}
+		if err := client.SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			logger.Errorf("failed to set pool size. %+v", err)
```
```suggestion
			logger.Errorf("failed to set pool size. %v", err)
```

leseb commented on pkg/operator/ceph/file/controller.go:
```diff
@@ -225,6 +225,19 @@ func (c *FilesystemController) filesystemOwner(fs *cephv1.CephFilesystem) metav1
 
 func filesystemChanged(oldFS, newFS cephv1.FilesystemSpec) bool {
+	if oldFS.MetadataPool.Replicated.Size != newFS.MetadataPool.Replicated.Size {
+		logger.Infof("number of MetaDataPool changed from %d to %d", oldFS.MetadataPool.Replicated.Size, newFS.MetadataPool.Replicated.Size)
```
```suggestion
		logger.Infof("replica size of MetaDataPool changed from %d to %d", oldFS.MetadataPool.Replicated.Size, newFS.MetadataPool.Replicated.Size)
```

rajatsingh25aug commented:
Integration Tests were failed the last time, need to check that if that happened again!

rajatsingh25aug commented:
It didn't this time.

leseb commented:
@travisn PTAL

rajatsingh25aug commented:
> @rajatsingh25aug is it still WIP? If not please update the title of your PR.

Done!.
Thanks

rajatsingh25aug commented:
> @rajatsingh25aug The code LGTM. Could you just verify a few more scenarios if you haven't already?
> 
> 1. Install the filesystem with replica 1
Not sure what replica are you talking about if is it the pools replicas?. 

> 2. Increase the replicas to 2. Verify that the setting is applied.
Changed the replica for both the pools.
Also, I changed the MDS activeCount. I noticed no errors.

> 3. Restart the operator. Verify that there are no errors in the log from setting the size again on the full orchestration.
No errors from the operator.

> 4. Decrease the replicas back to 1. Verify that the setting is applied.
Settings Applied without any problems.

The only thing is that there are 4 pods for `rook-ceph-mds` are running namely `rook-ceph-mds-myfs-a`  up to `rook-ceph-mds-myfs-d` and on my 3 master-worker setups, one out of four pods i.e `rook-ceph-mds-myfs-d` is not able to get scheduled on any node and hence was in `pending` state.
That because of toleration is not set for this pod.

Apart from this, everything is running smoothly and I was able to perform your operations without any issues.

rajatsingh25aug commented:
Strange that my test cases are failing here but not locally.

rajatsingh25aug commented:
@travisn  can you please review this?.
It's ready to review but the CI is failing but again, the tests are passing locally from my end.
Thanks

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -93,6 +96,19 @@ func ListFilesystems(context *clusterd.Context, clusterName string) ([]CephFiles
 	return filesystems, nil
 }
 
+func GenerateMetadataPoolName(fsName string, metadataPool *model.Pool) *model.Pool {
```
exported function GenerateMetadataPoolName should have comment or be unexported

leseb replied:
the comment must start by the name of the function.

rajatsingh25aug replied:
I have changed the comments and now they start with the name of the functions and also I have added the commit message describing what my PR is actually doing.
Thanks again Seb!

leseb replied:
It's not a method but a function since it has no receiver

BlaineEXE replied:
This should have a fuller description as well. In my words, this is a helper method that generates a new name for the given metadata pool and returns the renamed pool.

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -93,6 +96,19 @@ func ListFilesystems(context *clusterd.Context, clusterName string) ([]CephFiles
+
+}
+
+func GenerateDataPoolName(fsName string, dataPools []*model.Pool) []*model.Pool {
```
ditto

BlaineEXE replied:
This should have a fuller description as well. In my words, this is a helper method that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list. It also should end in an `s` (`GenerateDataPoolNames`) because it can take an array of inputs, not just a single pool.

leseb commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -174,6 +190,26 @@ func IsMultiFSEnabled() bool {
 	}
 	return false
 }
+func SetPoolSize(context *clusterd.Context, clusterName, fsName string, dataPool []*model.Pool, metadataPool *model.Pool) error {
```
ditto

BlaineEXE commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -175,6 +193,28 @@ func IsMultiFSEnabled() bool {
+	metadataPool = GenerateMetadataPoolName(fsName, metadataPool)
+	err := SetPoolProperty(context, clusterName, metadataPool.Name, "size", strconv.Itoa(int(metadataPool.ReplicatedConfig.Size)))
+	if err != nil {
+		return errors.Wrapf(err, "failed to update metadataPool replica for filesystem %q", fsName)
```
I think errors here should give the user information about what is wrong and how badly it is wrong. I would suggest that these messages include both the replica count that Rook tried to apply but failed, and upon failure we should try to see what the current replica count is to also report so that users can decide how badly their data safety is currently different from what they desire.

rajatsingh25aug replied:
Ok, so in the existing error Information, I will add something like
`failed to update metadataPool replica for filesystem %q, rook tried to change the replicaCount  from %d to %d.`
How is this?.

BlaineEXE replied:
It can be more succinct `failed to update metadataPool replica for filesystem %q from %d to %d`

BlaineEXE commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -176,6 +169,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 			)
 		}
+		if err := client.SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			logger.Errorf("failed to set pool size. %v", err)
```
@travisn @leseb do you think this should be a logged error or if this should return an error from the function? I'm on the fence. This could mean that the user's data is not as safe as they wish it to be if the pool size is increasing, or it could be less impactful if the size is decreasing. How do we expect that users will see logged error messages that don't cause overt errors? Do we assume they use a log monitor?

rajatsingh25aug commented on pkg/daemon/ceph/client/filesystem.go:
```diff
@@ -175,6 +193,28 @@ func IsMultiFSEnabled() bool {
+		if pool == nil {
+			continue
+		}
+		err = SetPoolProperty(context, clusterName, pool.Name, "size", strconv.Itoa(int(pool.ReplicatedConfig.Size)))
```
> Something doesn't seem right if we're adding new code for setting the pool size property. For example:
> 
> * Erasure coded pools wouldn't be handled

So I have to take care of the Erasure coded pools too, noted.
> * There is already code for creating pools for a new filesystem. We should be able to trigger than code in the orchestration instead of writing new code. Do you see where the pools are configured for a new filesystem?

Instead of writing a new function, just modify the existing function where pools were getting created?.

rajatsingh25aug replied:
> Something doesn't seem right if we're adding new code for setting the pool size property. For example:
> 
> * Erasure coded pools wouldn't be handled
I didn't get this, how am I ignoring the EC pools, say I have 10 pools which are both EC and non-EC. If they are being passed through a function to change their pool size, what am I missing in it?.
> * There is already code for creating pools for a new filesystem. We should be able to trigger than code in the orchestration instead of writing new code. Do you see where the pools are configured for a new filesystem?


JHel

rajatsingh25aug replied:
Hey Travis, I didn't get that, how am I missing the EC pools exactly, If there's a list of pools, how am I missing any of them regardless of they are EC or not.

cc @travisn  @leseb

rajatsingh25aug commented:
and what about the functions that are setting the names, are those needs to exist or not?
thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -160,13 +156,35 @@ func newFS(name string, metadataPool *model.Pool, dataPools []*model.Pool, activ
+func SetPoolSize(context *clusterd.Context, clusterName, fsName string, dataPool []*model.Pool, metadataPool *model.Pool) error {
+	// updating the MetaDataPool size
+	metadataPool = GenerateMetadataPoolName(fsName, metadataPool)
+	newPool := client.ModelPoolToCephPool(*metadataPool)
```
thanks @travisn I made some changes but still not sure if they contain all the cases that we need.
I'm still confused if I call the full `CreatePoolWithProfile()` method  instead of `SetPoolSize()` OR
 implement `SetPoolSize()` method in the similar pattern as `CreatePoolWithProfile()`
Please review it and let me know what am I missing.
Thanks

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -249,3 +271,18 @@ func downFilesystem(context *clusterd.Context, clusterName, filesystemName strin
 }
+
+// GenerateMetadataPoolName is a helper function that generates a new name for the given metadata pool and returns the renamed pool.
+func GenerateMetadataPoolName(fsName string, metadataPool *model.Pool) *model.Pool {
```
Where are these 2 new functions exported?

rajatsingh25aug replied:
Because they were being used by the functions `newFS` and `SetPoolSize`, but I have changed their names to lowercase as they are only being used `rook/rook/pkg/operator/ceph/file/filesystem.go`.
Thanks for the alert.

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -249,3 +271,18 @@ func downFilesystem(context *clusterd.Context, clusterName, filesystemName strin
+// GenerateDataPoolNames is a helper method that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
+func GenerateDataPoolNames(fsName string, dataPools []*model.Pool) []*model.Pool {
+	for i, pool := range dataPools {
+		pool.Name = fmt.Sprintf("%s-%s%d", fsName, dataPoolSuffix, i)
```
```suggestion
		pool.Name = fmt.Sprintf("%s-%s-%d", fsName, dataPoolSuffix, i)
```

Don't we want another separator for clarity?

rajatsingh25aug commented:
Thanks @travisn 
I tried your steps.

-  I was able to create the ecpools, changed the replica size for the other pools (metadata and data pools).

- I was also able to delete the operator and it was able to pick the changes.

- Upon changing the values (from the YAML and then doing `oc apply -f filename.yaml`)

Operator logs were like this and values were changes when I verified that from the ceph toolbox.
```
2020-03-24 12:05:49.129139 I | ceph-spec: CR has changed for "myfs". diff=  v1.FilesystemSpec{
  	MetadataPool: v1.PoolSpec{
  		FailureDomain: "",
  		CrushRoot:     "",
  		DeviceClass:   "",
  		Replicated: v1.ReplicatedSpec{
- 			Size:                   0x06,
+ 			Size:                   0x02,
  			TargetSizeRatio:        0,
  			RequireSafeReplicaSize: true,
  		},
  		ErasureCoded: v1.ErasureCodedSpec{},
  	},
  	DataPools: []v1.PoolSpec{
  		{
  			FailureDomain: "host",
  			CrushRoot:     "",
  			DeviceClass:   "",
  			Replicated: v1.ReplicatedSpec{
- 				Size:                   0x06,
+ 				Size:                   0x02,
  				TargetSizeRatio:        0,
  				RequireSafeReplicaSize: true,
  			},
  			ErasureCoded: v1.ErasureCodedSpec{},
  		},
  	},
  	PreservePoolsOnDelete: true,
  	MetadataServer:        v1.MetadataServerSpec{ActiveCount: 1, ActiveStandby: true, Placement: v1.Placement{PodAntiAffinity: &v1.PodAntiAffinity{RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{{LabelSelector: &v1.LabelSelector{MatchExpressions: []v1.LabelSelectorRequirement{{Key: "app", Operator: "In", Values: []string{"rook-ceph-mds"}}}}, TopologyKey: "kubernetes.io/hostname"}}, PreferredDuringSchedulingIgnoredDuringExecution: []v1.WeightedPodAffinityTerm{{Weight: 100, PodAffinityTerm: v1.PodAffinityTerm{LabelSelector: &v1.LabelSelector{MatchExpressions: []v1.LabelSelectorRequirement{{Key: "app", Operator: "In", Values: []string{"rook-ceph-mds"}}}}, TopologyKey: "topology.kubernetes.io/zone"}}}}}},
  }
```

rajatsingh25aug commented:
This was the pools info, notice the ecpool that I just created
```
pool 1 'myfs-metadata' replicated size 2 min_size 2 crush_rule 1 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode warn last_change 27 flags hashpspool,creating stripe_width 0 pg_autoscale_bias 4 pg_num_min 16 recovery_priority 5 application cephfs
pool 2 'myfs-data0' replicated size 2 min_size 2 crush_rule 2 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode warn last_change 29 flags hashpspool,creating stripe_width 0 application cephfs
pool 3 'ecpool' erasure size 3 min_size 2 crush_rule 3 object_hash rjenkins pg_num 64 pgp_num 64 autoscale_mode warn last_change 17 flags hashpspool,creating stripe_width 8192
```

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -175,6 +193,10 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 		}
+		if err := SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			return errors.Wrapf(err, "failed to set pool size to %d", f.metadataPool.ReplicatedConfig.Size)
+
```
Done

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -33,9 +33,9 @@ import (
-	dataPoolSuffix     = "data"
-	metadataPoolSuffix = "metadata"
 	appName            = "cephfs"
+	metaDataPoolSuffix = "-metadata"
```
removed, yes I observed that it was having double dashes.
Thanks.

rajatsingh25aug commented on cluster/examples/kubernetes/ceph/operator-openshift.yaml:
```diff
@@ -145,240 +145,240 @@ spec:
     spec:
       serviceAccountName: rook-ceph-system
       containers:
-      - name: rook-ceph-operator
```
no, my bad that I didn't made changes to the yaml after testing it.

rajatsingh25aug replied:
My vscode is adding extra quotes, I have to look into it.

rajatsingh25aug replied:
Done!, please review now, I have fixed all this.
thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -160,13 +156,31 @@ func newFS(name string, metadataPool *model.Pool, dataPools []*model.Pool, activ
+	metadataPool = generateMetadataPoolName(fsName, metadataPool)
+	err := client.CreatePoolWithProfile(context, clusterName, *metadataPool, appName)
+	if err != nil {
+		errors.Wrapf(err, "failed to update metadata pool %q", *&metadataPool.Name)
```
I took the reference from here
https://github.com/rook/rook/blob/master/pkg/operator/ceph/pool/controller.go#L241
It will print the stack trace I believe.
What do you think?.

rajatsingh25aug replied:
Fixed this too, thanks.

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -160,13 +156,31 @@ func newFS(name string, metadataPool *model.Pool, dataPools []*model.Pool, activ
+	for _, pool := range dataPool {
+		err := client.CreatePoolWithProfile(context, clusterName, *pool, appName)
+		if err != nil {
+			errors.Wrapf(err, "failed to update metadata pool %q", *&pool.Name)
```
Thanks, I missed it.

rajatsingh25aug commented on cluster/examples/kubernetes/ceph/filesystem.yaml:
```diff
@@ -13,13 +13,13 @@ spec:
   metadataPool:
     replicated:
-      size: 3
+      size: 6
```
Sure.

rajatsingh25aug replied:
Done!, please review now, I have fixed all this.
thanks

rajatsingh25aug commented on cluster/examples/kubernetes/ceph/operator-openshift.yaml:
```diff
@@ -60,7 +60,7 @@ allowPrivilegedContainer: true
 allowHostDirVolumePlugin: true
 priority:
-allowedCapabilities: ['*']
+allowedCapabilities: ["*"]
```
No, I dont know why this got changed.
My VScode sometime acts strange.

rajatsingh25aug replied:
Done!, please review now, I have fixed all this.
thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -175,6 +189,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 			)
 		}
+		if err := SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			return errors.Wrapf(err, "failed to set pool size to %d", f.metadataPool.ReplicatedConfig.Size)
```
Wrapped error?. I edited it and made changes like this
`return errors.Wrapf(err, "failed to set pools size")`

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,6 +141,25 @@ func newFS(name, namespace string) *Filesystem {
 	}
 }
 
+// SetPoolSize method sets the sizes for MetadataPool and dataPool
```
It's not a method but a function.

leseb replied:
not addressed

rajatsingh25aug replied:
My bad, Check now. Thanks

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,6 +141,25 @@ func newFS(name, namespace string) *Filesystem {
+	metadataPool = generateMetadataPoolName(fsName, metadataPool)
+	err := client.CreatePoolWithProfile(context, clusterName, *metadataPool, appName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update metadata pool %q", *&metadataPool.Name)
```
`*&metadataPool.Name` ???

rajatsingh25aug replied:
I was using this to point the Pool's name, strange that after reading, it is not reaching to Pool's name anymore.

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,6 +141,25 @@ func newFS(name, namespace string) *Filesystem {
+	for _, pool := range dataPool {
+		err := client.CreatePoolWithProfile(context, clusterName, *pool, appName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update datapool  %q", *&pool.Name)
```
`*&pool.Name` ???

rajatsingh25aug replied:
ditto

leseb replied:
At this point `pool` is just a string AFAIK so why that construct?

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -156,6 +176,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 			)
 		}
+		if err := SetPoolSize(context, clusterName, f.Name, f.dataPools, f.metadataPool); err != nil {
+			return errors.Wrapf(err, "failed to set pools size")
```
```suggestion
			return errors.Wrap(err, "failed to set pools size")
```

rajatsingh25aug replied:
Thanks, since there is no format specifier, `Wrap` will be used.

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +255,17 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
 }
+
+// generateMetadataPoolName is a helper function that generates a new name for the given metadata pool and returns the renamed pool.
+func generateMetadataPoolName(fsName string, metadataPool *model.Pool) *model.Pool {
```
Model pool is gone, we must rebase and adapt your code.

rajatsingh25aug replied:
Yes, and that is why I believe that I'm not able to use `*&pool.Name` anymore.
But what kind of changes needs to be adapted?.

leseb replied:
You just need to use values from the spec.

rajatsingh25aug replied:
Ok, I looked into this and I think from the spec you meant `cephv1.FilesystemSpec`.
I made changes according to that.
Let me know if I missed anything.
thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -249,3 +271,18 @@ func downFilesystem(context *clusterd.Context, clusterName, filesystemName strin
 }
+
+// GenerateMetadataPoolName is a helper function that generates a new name for the given metadata pool and returns the renamed pool.
+func GenerateMetadataPoolName(fsName string, metadataPool *model.Pool) *model.Pool {
```
I have done that, please check.
Thanks

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -249,3 +271,18 @@ func downFilesystem(context *clusterd.Context, clusterName, filesystemName strin
+}
+
+// GenerateDataPoolNames is a helper method that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
+func GenerateDataPoolNames(fsName string, dataPools []*model.Pool) []*model.Pool {
```
Done

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,9 +140,27 @@ func newFS(name, namespace string) *Filesystem {
 }
 
+// SetPoolSize method sets the sizes for MetadataPool and dataPool
+func SetPoolSize(f *Filesystem, context *clusterd.Context, clusterName, fsName string, spec cephv1.FilesystemSpec) error {
```
Since you have `f *Filesystem`, you don't need `clusterName, fsName string`

rajatsingh25aug replied:
Make sense, because I think I can use `f.Namespace` and `f.Name`  instead of `clusterName, fsName string`

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,9 +140,27 @@ func newFS(name, namespace string) *Filesystem {
+	for _, pool := range dataPoolsName {
+		err := client.CreatePoolWithProfile(context, f.Namespace, pool, spec.DataPools, appName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update datapool  %s", pool)
```
```suggestion
			return errors.Wrapf(err, "failed to update datapool %q", pool)
```

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +253,13 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
 	return nil
 }
+
+// generateDataPoolNames is a helper method that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
```
```suggestion
// generateDataPoolNames is a helper function that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
```

leseb commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +253,13 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
 }
+
+// generateDataPoolNames is a helper method that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
+func generateDataPoolNames(f *Filesystem, fsName string, spec cephv1.FilesystemSpec) []string {
```
You don't need `fsName` since you have `Filesystem`, you are not even using it.

rajatsingh25aug commented:
This commit should fix all the issues that you guys mentioned. Thanks a lot, seems like I need to learn a lot.

cc @travisn  @leseb

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,9 +140,27 @@ func newFS(name, namespace string) *Filesystem {
+// SetPoolSize function sets the sizes for MetadataPool and dataPool
+func SetPoolSize(f *Filesystem, context *clusterd.Context, spec cephv1.FilesystemSpec) error {
+	// generating the metadata pool's name
+	metadataPoolName := fmt.Sprintf("%s-%s", f.Name, metadataPoolSuffix)
```
Yes, I refactored it in the function down below.

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +255,19 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
+
+// generateMetaDataPoolNames is a function that generates a name for the MetaDataPool by prefixing the FileSystem's Name to the  constatnt metaDataPoolSuffix
+func generateMetaDataPoolName(f *Filesystem) string {
+	metaDataPoolName := fmt.Sprintf("%s-%s", f.Name, metaDataPoolSuffix)
```
Done.

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -193,10 +216,10 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 
-	var dataPoolNames []string
+	var dataPoolsName []string
+	dataPoolsName = generateDataPoolNames(f, spec)
```
Sure

rajatsingh25aug commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -140,6 +140,26 @@ func newFS(name, namespace string) *Filesystem {
+		return errors.Wrapf(err, "failed to update metadata pool %q", metadataPoolName)
+	}
+	// generating the data pool's name
+	dataPoolsName := generateDataPoolNames(f, spec)
```
Done

rajatsingh25aug commented:
Hi, I tried looking into some of the `log` files but  am I supposed to look at `_output/tests/v1.16.8_integrationTests.log`  or `_output/tests/CephFlexDriverSuite_TestFileSystem_aws_1.14.x_flex-ns-system_rook-ceph-operator-5bc4f59967-87ppf_1585662848.log` 
Also, I see that there are lots of files like these, do I have to go to all of them?.

sp98 commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -193,10 +216,10 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 
-	var dataPoolNames []string
+	var dataPoolsName []string
+	dataPoolNames := generateDataPoolNames(f, spec)
```
A whole new variable `dataPoolNames` is created here. So now we have two variables `dataPoolNames` and 'dataPoolsName`.  And `dataPoolsName` is always empty.  This looks like the reason for integration tests failure.

rajatsingh25aug replied:
Done, thanks

sp98 commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +255,18 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
 	return nil
 }
+
+// generateDataPoolNames is a function that generates a new name for each pool in the input list suffixed with an appropriate index number, and it returns the renamed pool list
```
NIT: No need to write `is a function`.  Same with other method descriptions as well.

rajatsingh25aug replied:
I wrote this after discussing with Sebastian. So I don't think this is a problem.

sp98 commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -232,3 +254,18 @@ func downFilesystem(context *clusterd.Context, namespace, filesystemName string)
+	return dataPoolNames
+}
+
+// generateMetaDataPoolNames is a function that generates a name for the MetaDataPool by prefixing the FileSystem's Name to the  constatnt metaDataPoolSuffix
```
```suggestion
// generateMetaDataPoolName generates MetaDataPool name by prefixing the filesystem name to the constant metaDataPoolSuffix
```

rajatsingh25aug replied:
done

BlaineEXE commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -156,6 +176,9 @@ func (f *Filesystem) doFilesystemCreate(context *clusterd.Context, cephVersion c
 			)
 		}
+		if err := SetPoolSize(f, context, spec); err != nil {
+			return errors.Wrap(err, "failed to set pools size")
```
Should this report the filesystem name for users?

rajatsingh25aug replied:
it just reports that whether or not the pool size were able to set or not?.
